feat(insights): Compare to arbitrary prior periods of time

aspicer opened the pull request:
## Problem
The "compare to previous period" function currently only allows you to compare to the period immediately preceding what is shown on the graph. Resolves #14991.

## Changes

Change the compare selector to let the user select an arbitrary time in the past to compare with

Works with trends and stickiness

Known bugs for later:
[ ] If a breakdown exists in a previous query but not in the current one, it won't show up in the selector

<img width="726" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/e2c68811-84fb-473d-8620-d73c362c8697">

<img width="1019" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/f86032fc-4b72-4e3f-b02a-44307e149381">

<img width="1025" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/37cdde8d-c13b-4a46-99c1-7992e818c410">

<img width="748" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/f65ba313-c8f5-4bd7-a828-80b58625d580">

The old functionality is still available as "period":
<img width="1018" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/328173f4-dada-4f1d-8441-b09e28233658">


## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Tested in dev. Wrote tests for the python date class that handles the logic.

thmsobrmlr reviewed (commented):
Cool stuff!

Thought I'd leave you an early hint that this needs to be handled in actors queries and tooltips as well. Some more integrated tests would be great as well :)

*Edit:*

**Actors modal**

<img width="611" alt="Screenshot 2024-05-21 at 22 41 31" src="https://github.com/PostHog/posthog/assets/1851359/0e61e9b1-9c1e-4cdb-bc2e-4a069ec6e3f6">

**Tooltips**
<img width="738" alt="Screenshot 2024-05-21 at 22 41 47" src="https://github.com/PostHog/posthog/assets/1851359/a7b6eef7-968e-4cf4-8c12-301a5c0002a5">

aspicer commented:
@thmsobrmlr Thanks for the heads up! I made those changes and wrote more thorough tests for the trends query, actors query, and stickiness query.

I saw your PR here with notes about future fixes for actors: https://github.com/PostHog/posthog/pull/22132
I'm happy to help chip away at any outstanding issues - let's sync up tomorrow at the standup or afterwards on what issues I could help resolve.

As for this PR, here's an example of the actors query working. The insight below has a 2 day compare set up.
You can then see that the actors query for `current May 15th` returns the same data as `previous May 17th`
<img width="1019" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/6d723ed1-eb23-41f7-8124-56fbac4540b6">
<img width="574" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/8dcb5ef1-f28a-41d4-98eb-b379acffcbb1">
<img width="567" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/3c7c465c-6892-4c75-a2e4-7180b5be8fdf">

mariusandra commented:
It looks like there's another PR out there implementing almost the same stuff: https://github.com/PostHog/posthog/pull/22267

Twixes commented on frontend/src/lib/components/CompareFilter/CompareFilter.tsx:
```diff
@@ -16,13 +17,28 @@ export function CompareFilter(): JSX.Element | null {
 
+    const label = (
+        <span className="font-normal">
+            <RollingDateRangeFilter
+                dateRangeFilterLabel="Compare to "
+                dateRangeFilterSuffixLabel=" earlier"
+                allowPeriod={true}
+                dateFrom={compareTo}
+                inUse={true}
+                onChange={(compareTo) => {
+                    updateInsightFilter({ compare: true, compareTo })
+                }}
+            />
```
This current design doesn't fit in super well into the existing UI, because:
1. `CompareFilter` is currently taller than any other elements in its row
2. This `RollingDateRangeFilter` is rendered as a button, but it shouldn't. Should be just a div really, and only the select for choosing the comparison unit should be a button

But with the above addressed it should be pretty sleek!

aspicer replied:
<img width="1018" alt="image" src="https://github.com/PostHog/posthog/assets/1855120/8c68c6e9-69d0-43ed-91b2-9d553fb555d6">

Is this what you were imagining?

aspicer replied:
I like it. It's very clear. I wasn't happy with compare to period earlier anyways.

aspicer replied:
https://github.com/PostHog/posthog/assets/1855120/245a3817-b1d7-4c98-8757-45cf49ec15c2

let me know what you think of this

Twixes replied:
This is pretty great

Twixes commented on frontend/src/scenes/insights/utils/cleanFilters.ts:
```diff
@@ -465,6 +465,7 @@ export function cleanFilters(
 
         if (filters.date_from === 'all' || isLifecycleFilter(filters)) {
             trendLikeFilter['compare'] = false
+            trendLikeFilter['compare_to'] = undefined
```
A somewhat radical idea I have is to turn `compare` into this type: `bool | { relativePeriodStart: string }` – in which case it would be _impossible_ to have `compare_to` set while `compare` is false. But I'm not sure if it'd actually work smoothly. @thmsobrmlr What do you think, having worked out a lot of the schema of queries?

thmsobrmlr replied:
I don't see why this wouldn't work, but I'm also not sure if this would make it more understandable. Another option would be to use a discriminating union (frontend side, don't think python supports somthing like it) `{compare:true, relativePeriodStart: null | string} | {compare:false}`

Aside: One thing we're trying to do with the new schema is grouping related properties together in a filter e.g." `breakdownFilter`. In line with that we could have a `compareFilter` here that has all options for the compare feature.

Twixes commented on posthog/hogql_queries/insights/stickiness_query_runner.py:
```diff
@@ -376,7 +380,9 @@ def setup_series(self) -> list[SeriesWithExtras]:
     def date_range(self, series: SeriesWithExtras):
         if series.is_previous_period_series:
-            return self.query_previous_date_range
+            if not series.compare_to:
+                return self.query_previous_date_range
```
I would say this `if` would be better encapsulated _within_ `query_compare_to_date_range()`

aspicer replied:
Yes, much cleaner.
![image](https://github.com/PostHog/posthog/assets/1855120/7495eda4-b47e-4897-9b04-026640d65736)

Twixes reviewed (changes requested):
Really useful, just the UI to simplify

Twixes commented on frontend/src/lib/components/CompareFilter/CompareFilter.tsx:
```diff
@@ -1,30 +1,117 @@
 
+    // This keeps the state of the rolling date range filter, even when different drop down options are selected
+    // The default value for this is one month
+    const [compareToUi, setCompareToUi] = useState<string>(compareTo || '-1m')
```
Absolute nit: I think more conventionally/clearer for us this would be `tentativeCompareTo`

Twixes commented on frontend/src/lib/components/CompareFilter/CompareFilter.tsx:
```diff
@@ -1,30 +1,117 @@
         return null
     }
 
+    if (!!compareTo && compareToUi != compareTo) {
+        setCompareToUi(compareTo)
+    }
```
This should be inside a `useEffect()`

Twixes commented on frontend/src/lib/components/CompareFilter/CompareFilter.tsx:
```diff
@@ -1,30 +1,117 @@
+        {
+            value: 'compareTo',
+            label: (
+                <div onClick={() => updateInsightFilter({ compare: true, compareTo: compareToUi })}>
```
Seems non-typical for an `onClick` to be on a div, maybe `LemonSelect`'s `onSelect` prop could do this job here?

Twixes commented on frontend/src/lib/components/CompareFilter/CompareFilter.tsx:
```diff
@@ -1,30 +1,117 @@
+        />
+    )
+
+    /*
+    const label = (
```
Commented out?

Twixes commented on frontend/src/lib/components/DateFilter/RollingDateRangeFilter.tsx:
```diff
@@ -17,8 +17,12 @@ const dateOptions: LemonSelectOptionLeaf<DateOption>[] = [
 type RollingDateRangeFilterProps = {
+    isButton?: boolean
     pageKey?: string
+    // specifies if the filter is selected in the dropdown (to darken)
```
As a VS Code user I request using the JSDoc notation (`/** */`), because that gets picked up by VSC for hover tooltips (I think JetBrains is more lax when it comes to JS – I see the point in JSDoc, because it means documentation is separate from notes)

Twixes reviewed (commented):
One last round

Twixes commented:
Is this up for re-review? @aspicer Lemme know with a review request

aspicer commented:
@thmsobrmlr @Twixes can i get a new look at this PR? it should be good to go with design and backend changes.

I have
- Replaced the UI with @Twixes's dropdown selection.
- Moved compare and compare_to onto a compareFilter and refactored accordingly
- Updated web analytics to use the new format
- Changed the front-end "all time" selector to remove compare, which the backend already does when you save
- Updated tests to include the new format

I have tested this in dev and also tested to ensure that current saved insights keep their "compare" flag when you switch from master to this branch.

Thanks, let me know if you have questions

https://github.com/PostHog/posthog/assets/1855120/fb890186-8a6d-4fb7-b893-7e4628c3795b

thmsobrmlr commented on frontend/src/scenes/notebooks/Notebook/migrations/migrate.test.ts:
```diff
@@ -13,7 +13,7 @@ describe('migrate()', () => {
                     type: 'ph-query',
                     attrs: {
-                        query: '{"kind":"InsightVizNode","source":{"kind":"TrendsQuery","properties":{"type":"AND","values":[{"type":"AND","values":[]}]},"filterTestAccounts":true,"dateRange":{"date_to":null,"date_from":"-90d"},"series":[{"kind":"EventsNode","event":"$pageview","name":"$pageview","properties":[{"key":"$referring_domain","type":"event","value":"google|duckduckgo|brave|bing","opera ...
+                        query: '{"kind":"InsightVizNode","source":{"kind":"TrendsQuery","properties":{"type":"AND","values":[{"type":"AND","values":[]}]},"filterTestAccounts":true,"dateRange":{"date_to":null,"date_from":"-90d"},"series":[{"kind":"EventsNode","event":"$pageview","name":"$pageview","properties":[{"key":"$referring_domain","type":"event","value":"google|duckduckgo|brave|bing","opera ...
```
Didn't get around to a full review yet, so just a fly-by that I think we need to actually adapt the `migrate.ts` function for notebook nodes with queries to still work.

aspicer replied:
Thanks for the heads up. I wasn't familiar with notebook functionality.

I added the migration code here:
https://github.com/PostHog/posthog/pull/22397/files#diff-fef986adb6581e65a4f5a219af0c0aba9896c6e22ab650ee3912843ff761ff4a

Twixes reviewed (approved):
Looks great to me, although I'll defer to @thmsobrmlr on the query schema

thmsobrmlr reviewed (changes requested):
Very solid work overall, @aspicer!

Did a deep dive and found two things:
1. When opening the actors modal for an insight with compare enabled, the dropdown for compare options doesn't always show any more. These dropdowns are rendered dynamically based on an `InsightActorsQueryOptions` query to the backend, so I suspect this needs some adaptations as well.
2. The migration for notebooks does run on already migrated nodes as well. One of the migrations in there migrates snake-cased keys to camel cased ones e.g. `aggregation_axis_format` to `aggregationAxisFormat`. These insights can still have a `compare` key in `trendsFilter`, and since this key was added to `legacyKeys`, the insight will be fully migrated again. Now the already migrated snake-cased keys don't exist any more, resulting in an empty `trendsFilter`. I added an failing test case to better show what I mean. This could possibly be solved by running the migrations in sequence.

aspicer commented:
@thmsobrmlr I fixed the issue you were describing, I think, for both trends and stickiness: https://github.com/PostHog/posthog/pull/22397/commits/30bdeacd241ae52cc7e7bbbe96a2512634c382c5#diff-fef986adb6581e65a4f5a219af0c0aba9896c6e22ab650ee3912843ff761ff4aR124

Hopefully this is what you meant?

As for the dropdown not showing up, I was only able to get it to repro once. It was because the async query for the dropdown values never returned. I didn't have any error logs in my celery or django terminals, so I wasn't able to really debug this. I think it's probably okay to ship? If I can repro it more consistently, I will fix it!

thmsobrmlr reviewed (approved):
The notebooks migration looks good and I can't replicate any issue with the actors modal options any more. Thanks for addressing! All great now ✨

aspicer committed:
code changes for compare

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)
